Cannot set mail.force_extra_parameters as PHP_INI_PERDIR

marcovtwout opened the pull request:
Documentation does not match: https://github.com/php/doc-en/blob/3cdd39bb9505e6735d802da83a04870cfa8f2311/appendices/ini.list.xml

This param cannot be set through PHP_INI_PERDIR because of https://bugs.php.net/bug.php?id=71901

marcovtwout commented on reference/mail/ini.xml:
```diff
@@ -31,7 +31,7 @@
      <entry><link linkend="ini.mail.force_extra_parameters">mail.force_extra_parameters</link></entry>
      <entry>NULL</entry>
-     <entry>PHP_INI_SYSTEM|PHP_INI_PERDIR</entry>
+     <entry>&php.ini; only</entry>
```
@cmb69 If httpd.conf is also allowed that means that https://github.com/php/doc-en/blob/3cdd39bb9505e6735d802da83a04870cfa8f2311/appendices/ini.list.xml is also incorrect.

And is PHP_INI_PERDIR minus .user.ini and .htaccess not effectively the same as PHP_INI_SYSTEM? (according to the options on https://www.php.net/manual/en/configuration.changes.modes.php)

cmb69 replied:
`PHP_INI_SYSTEM` only allows to set the value via `php_admin_value`, but `mail.force_extra_parameters` can be set via `php_value` in .httpconf. Frankly, I wonder why that setting wasn't simply changed to `PHP_INI_SYSTEM`, but we have to document the status quo.

marcovtwout replied:
@cmb69 Are you absolutely sure `mail.force_extra_parameters` can be set through the httpd config, and is not being reverted by `OnChangeMailForceExtra` specified in https://github.com/php/php-src/blob/PHP-7.1.1/main/main.c#L601 ?

cmb69 replied:
`OnChangeMailForceExtra` is only there to prevent modifications in .htaccess and .user.ini. Setting `mail.force_extra_parameters` has effect in .htconf via `php_value` (changes the local value; master value remains, as usual).

> Whatever the documented outcome should be, it must be applied to both files (ini.list.xml and ini.xml)

Indeed, although we should try to get rid of the duplication. If we won't care about the alphabetic order of the directives, the solution would be trivial. Otherwise, this is more ugly/brittle.

marcovtwout replied:
If I understand correctly this setting can be set in httpd-config through only with `php_value` and not with `php_admin_value`? Or can it be set either way?

Both documentation pages only allow for a very short description under column "Changeable". Therefor I would vote to document this in the most concise and clear way possible.

Some suggestions:
- `PHP_INI_SYSTEM` (if detail can be ommitted)
- `PHP_INI_SYSTEM with php_value`
- `php.ini, or httpd.conf with php_value`

Technical nuances could also be added under https://www.php.net/manual/en/mail.configuration.php#ini.mail.force_extra_parameters

afilina commented:
Pinging author: @marcovtwout
From my understanding, the reviewer's recommendations would make this PR acceptable.

cmb69 commented:
Hmm, as it is now, the PR would not change anything. Are you still working on that @marcovtwout? :)

afilina commented:
@cmb69 There has been no activity and it's still a zero-line PR. Does it make sense to close it?

marcovtwout commented:
@cmb69 I think we were almost there, we just have to pick up the discussion where we left off in https://github.com/php/doc-en/pull/1081/commits/03e618ddc08d59057c1eec29c512f9702172c469#r1060038293

If you could answer my remaining question there, I'll update the PR with one of the suggested solutions.

Girgias commented:
Make it ``PHP_INI_SYSTEM`` and add the full explanation in the descriptions, as this is probably something we should change/deprecate to make it just ``PHP_INI_SYSTEM`` in a future version (would maybe require an RFC)

afilina commented:
Closing and reopening to see what the pipeline is complaining about.

Girgias reviewed (changes requested):
This has no description of the nuance Christopher has said.

This should be documented in the ``mail/ini.xml`` file, under the relevant INI description section.

marcovtwout commented:
@Girgias You're right it doesn't, because I'm not entire sure what to put there exactly. My previous question is still awaiting reply:

> If I understand correctly this setting can be set in httpd-config through only with php_value and not with php_admin_value? Or can it be set either way?

marcovtwout commented:
Xmas time is a good time to revisit old PR's.

I tested behavior on current PHP version to verify how this ini setting can be changed. @cmb69 is correct that, in addition to normal INI_SYSTEM behavior, this value can also be set with `php_value` on the `httpd.conf` level. Setting this on the INI_PERDIR level (in either .htaccess or .user.ini) does not allow you to override the value. However, it has the interesting effect that if you try to do anyways (for example in .htaccess) the setting is **reverted** to the master value. That looks like a bug which should either be addressed or the entire special treatment of this setting removed in the future.

For now I would say that while you _can_ use `php_value`, it should not be recommended because of the buggy behavior or because it could be removed in the future. I rebased and updated the PR which makes the documentation consistent and documents the technical nuance.

cmb69 commented on reference/mail/ini.xml:
```diff
@@ -106,6 +106,11 @@
+    <para>
+     In addition to the default behavior for <constant>INI_SYSTEM</constant>,
+     this value can also be set with <literal>php_value</literal>
+     in <filename>httpd.conf</filename> (but this is not recommended)
```
```suggestion
     in <filename>httpd.conf</filename> (but this is not recommended).
```

cmb69 reviewed (approved):
Besides the minor nit, this looks good to me now. Thank you! And sorry for the long delay.

Girgias commented on reference/mail/ini.xml:
```diff
@@ -106,6 +106,11 @@
      parameters to the sendmail binary. These parameters will always replace
      the value of the 5th parameter to <function>mail</function>.
     </para>
+    <para>
+     In addition to the default behavior for <constant>INI_SYSTEM</constant>,
+     this value can also be set with <literal>php_value</literal>
+     in <filename>httpd.conf</filename> (but this is not recommended).
+    </para>
```
```suggestion
    <simpara>
     In addition to the default behavior for <constant>INI_SYSTEM</constant>,
     this value can also be set with <literal>php_value</literal>
     in <filename>httpd.conf</filename> (but this is not recommended).
    </simpara>
```
